client/src/components/pages/Players.js

```diff
@@ -1,289 +1,339 @@
+  const [loading, setLoading] = useState(false);
+  const [modalVisible, setModalVisible] = useState(false);
+  const [modalLoading, setModalLoading] = useState(false);
+  const [addPlayerData, setAddPlayerData] = useState({});
+  const [flags, setFlags] = useState(new Set());
+  const [editingTeam, setEditingTeam] = useState(-1);
+  const [rankRange, setRankRange] = useState(0);
+
+  const fetchPlayers = async () => {
+    try {
+      const [playersData, tourneyData] = await Promise.all([
+        get("/api/players", { tourney: tourney }),
+        get("/api/tournament", { tourney: tourney }),
+      ]);
+
+      setPlayers(playersData);
+      setHasTeams(tourneyData.teams);
+      setHasGroups(tourneyData.stages.some((s) => s.name === "Group Stage"));
+      setRankRange([
+        tourneyData.rankMin,
+        tourneyData.rankMax !== -1 ? tourneyData.rankMax : Infinity,
+      ]);
+      setFlags(new Set(tourneyData.flags || []));
+
+      if (tourneyData.teams) {
+        try {
+          const teamsData = await get("/api/teams", { tourney: tourney });
+          setTeams(teamsData);
+        } catch {
+          return message.error("Something went wrong, failed to fetch team data.");
+        }
+      }
+    } catch {
+      message.error("Something went wrong, failed to fetch data.");
     }
-  }
-
-  getStatsById = (_id) => {
-    const player = this.state.players.filter((p) => p._id === _id)[0];
-    if (player) return this.getStats(player);
-    return {};
   };
 
-  getStats = (player) => {
-    const stats = player.stats.filter((s) => s.tourney === this.props.tourney)[0];
+  useEffect(() => {
+    document.title = `${prettifyTourney(tourney)}: Players`;
+
+    fetchPlayers();
+  }, []);
+
+  const getStats = (player) => {
+    const stats = player.stats.filter((s) => s.tourney === tourney)[0];
     return stats || {};
   };
 
-  getRegTime = (player) => {
-    const regTime = this.getStats(player).regTime;
+  const getStatsById = (_id) => {
+    const player = players.filter((p) => p._id === _id)[0];
+    if (player) return getStats(player);
+    return {};
+  };
+
+  const getRegTime = (player) => {
+    const regTime = getStats(player).regTime;
     if (regTime) return moment(regTime).unix();
     return -1;
   };
 
-  // just sort on the frontend
-  sortPlayers(sort) {
-    this.setState((state) => {
-      const players = [...state.players];
+  const sortPlayers = (sortMethod, unsortedPlayers) => {
+    let playersData;
 
-      const getRank = (p) => p.rank || Infinity;
+    const getRank = (p) => p.rank || Infinity;
 
-      if (sort === "rank") {
-        players.sort((x, y) => getRank(x) - getRank(y));
-      } else if (sort === "seed") {
-        players.sort(
-          (x, y) =>
-            (this.getStats(x).seedNum || getRank(x)) - (this.getStats(y).seedNum || getRank(y))
+    switch (sortMethod) {
+      case "rank":
+        playersData = unsortedPlayers.sort((x, y) => getRank(x) - getRank(y));
```

Comment:
since `.sort` sorts the array in-place, this call mutates `unsortedPlayers` and thus is directly mutating the react state, which is generally frowned upon in react

instead, this function should create a copy of `unsortedPlayers` to avoid mutating the original array

example implementation that makes a copy with `[...unsortedPlayers]` and also uses an object to hold the sort functions to make it more concise/less repeated:
```
 const sortPlayers = (sortMethod, unsortedPlayers) => {
    const getRank = (p) => p.rank || Infinity;

    const sortFunctions = {
      rank: (x, y) => getRank(x) - getRank(y),
      seed: (x, y) => (getStats(x).seedNum || getRank(x)) - (getStats(y).seedNum || getRank(y)),
      group: (x, y) => ((getStats(x).group || "_") < (getStats(y).group || "_") ? -1 : 1),
      alpha: (x, y) => (x.username.toLowerCase() < y.username.toLowerCase() ? -1 : 1),
      country: (x, y) => (x.country < y.country ? -1 : 1),
      reg: (x, y) => getRegTime(x) - getRegTime(y),
    };

    const sortFn = sortFunctions[sortMethod];
    return [...unsortedPlayers].sort(sortFn);
  };
```